src/util.rs

```diff
@@ -52,34 +52,77 @@ pub fn ruffini<F: PrimeField>(
     DensePolynomial::from_coefficients_vec(quotient)
 }
 
-pub fn get_domain_attrs<F: FftField>(
-    domain: &GeneralEvaluationDomain<F>,
-    attr: &'static str,
-) -> F {
-    match domain {
-        GeneralEvaluationDomain::MixedRadix(domain) => match attr {
-            "group_gen" => domain.group_gen,
-            "group_gen_inv" => domain.group_gen_inv,
-            "generator_inv" => domain.generator_inv,
-            "size_as_fe" => domain.size_as_field_element,
-            "size_inv" => domain.size_inv,
-            _ => unreachable!(),
-        },
-        GeneralEvaluationDomain::Radix2(domain) => match attr {
-            "group_gen" => domain.group_gen,
-            "group_gen_inv" => domain.group_gen_inv,
-            "generator_inv" => domain.generator_inv,
-            "size_as_fe" => domain.size_as_field_element,
-            "size_inv" => domain.size_inv,
-            _ => unreachable!(),
-        },
-        _ => unreachable!(),
+/// Evaluation Domain Extension Trait
+pub trait EvaluationDomainExt<F>: EvaluationDomain<F>
+where
+    F: FftField,
+{
+    /// Returns the value of `log_2(self.size)`.
+    fn log_size_of_group(&self) -> u32;
+
+    /// Returns the inverse of the size in the field.
+    fn size_inv(&self) -> F;
+
+    /// Returns a fixed generator of the subgroup.
+    fn group_gen(&self) -> F;
+
+    /// Returns the inverse of the fixed generator of the subgroup.
+    fn group_gen_inv(&self) -> F;
+
+    /// Returns a fixed multiplicative generator of the finite field.
+    fn generator_inv(&self) -> F;
+}
+
+impl<F> EvaluationDomainExt<F> for GeneralEvaluationDomain<F>
+where
+    F: FftField,
+{
+    #[inline]
+    fn log_size_of_group(&self) -> u32 {
+        match self {
+            GeneralEvaluationDomain::Radix2(domain) => domain.log_size_of_group,
+            GeneralEvaluationDomain::MixedRadix(domain) => {
+                domain.log_size_of_group
+            }
+        }
+    }
+
+    #[inline]
+    fn size_inv(&self) -> F {
+        match self {
+            GeneralEvaluationDomain::Radix2(domain) => domain.size_inv,
+            GeneralEvaluationDomain::MixedRadix(domain) => domain.size_inv,
+        }
+    }
+
+    #[inline]
+    fn group_gen(&self) -> F {
+        match self {
+            GeneralEvaluationDomain::Radix2(domain) => domain.group_gen,
+            GeneralEvaluationDomain::MixedRadix(domain) => domain.group_gen,
+        }
+    }
+
+    #[inline]
+    fn group_gen_inv(&self) -> F {
+        match self {
+            GeneralEvaluationDomain::Radix2(domain) => domain.group_gen_inv,
+            GeneralEvaluationDomain::MixedRadix(domain) => domain.group_gen_inv,
+        }
+    }
+
+    #[inline]
+    fn generator_inv(&self) -> F {
+        match self {
+            GeneralEvaluationDomain::Radix2(domain) => domain.generator_inv,
+            GeneralEvaluationDomain::MixedRadix(domain) => domain.generator_inv,
+        }
```

Comment:
That is basically replaced by https://github.com/arkworks-rs/algebra/pull/333/files as soon as is released. So I won't merge it here. Specially considering the repo should ideally contain only PLONK-related logic and avoid adding any external stuff as much as possible (the solution is really good obviously) But I prefer to wait for the new version rather than have a lot of new code that we will need to remove soon anyway.

Reply:
Here are some reasons why this should be kept:
1. This repo is already containing non-PLONK logic. This is the nature of software and we cannot expect our code to be a spec implementation only.
2. Using `&'static str` to query members of an object cannot be checked by the compiler and so when we do something wrong like mistype the name we cannot know if it is wrong unless we have a test that passes through that line. Getting full code coverage is harder than just using extension `trait`s.
3. Using extension traits to extend non-local libraries is standard practice in Rust so other developers will recognize this design language and understand its purpose more readily.
4. Using a trait allows for better discoverability and documentation.
5. Using a trait allows for abstraction over implementations.

Reply:
After discussing with @davidnevadoc he agreed on leaving the struct/trait that you purposed. 

So fine from my side then.